api/envoy/extensions/quic/server_preferred_address/v3/fixed_server_preferred_address_config.proto

```diff
@@ -16,20 +18,51 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#extension: envoy.quic.server_preferred_address.fixed]
 
 // Configuration for FixedServerPreferredAddressConfig.
+// [#next-free-field: 7]
 message FixedServerPreferredAddressConfig {
   // [#comment:TODO(danzh2010): discuss with API shepherds before removing WiP status.]
 
   option (xds.annotations.v3.message_status).work_in_progress = true;
 
   oneof ipv4_type {
     // String representation of IPv4 address, i.e. "127.0.0.2".
+    // The listener's port will be used.
     // If not specified, none will be configured.
     string ipv4_address = 1;
+
+    // The IPv4 address to advertise to clients for Server Preferred Address.
+    config.core.v3.SocketAddress ipv4_address_and_port = 3;
+  }
+
+  oneof ipv4_dnat_type {
+    // If there is a DNAT between the client and Envoy, the address that Envoy will observe
+    // server preferred address IPv4 packets being sent to. If this is not specified, it is assumed
+    // there is no DNAT and the server preferred address packets will be sent to the address advertised
+    // to clients for server preferred address.
+    //
+    // String representation of IPv4 address, i.e. "127.0.0.2"
+    // The listener's port will be used.
+    string ipv4_dnat_address = 5;
   }
 
   oneof ipv6_type {
     // String representation of IPv6 address, i.e. "::1".
+    // The listener's port will be used.
     // If not specified, none will be configured.
     string ipv6_address = 2;
+
+    // The IPv6 address to advertise to clients for Server Preferred Address.
+    config.core.v3.SocketAddress ipv6_address_and_port = 4;
+  }
+
+  oneof ipv6_dnat_type {
+    // If there is a DNAT between the client and Envoy, the address that Envoy will observe
+    // server preferred address IPv6 packets being sent to. If this is not specified, it is assumed
+    // there is no DNAT and the server preferred address packets will be sent to the address advertised
+    // to clients for server preferred address.
+    //
+    // String representation of IPv6 address, i.e. "::1"
+    // The listener's port will be used.
+    string ipv6_dnat_address = 6;
```

Comment:
`oneof` is now discouraged: https://github.com/envoyproxy/envoy/blob/21b406b43e89ffcd92ceab55a462c777e07a3711/api/STYLE.md?plain=1#L63

Is it necessary in this case?  Do we anticipate more fields to select from?

Reply:
I'm trying to mirror the other field here. It'll probably need a similar set of options as `ipv6_address` is getting in this PR. See https://github.com/envoyproxy/envoy/pull/33774#discussion_r1578657986 for discussion on why it is this way.

Reply:
Thinking about this a little further, I think using the config.core.v3.Address is a better type for this data, and in the cases where we currently cannot specify a port, ignore it (or throw a validation error if it isn't zero). Then we can deprecate the existing string version, and stop using the oneof's at all. Wdyt?

Reply:
@abeyad and @RyanTheOptimist would it be better like this (plus the equivalent for ipv6)?

```
  oneof ipv4_type {
    // Deprecated
    string ipv4_address = 1;

  }

  // The IPv4 address to advertise to clients for Server Preferred Address.
  config.core.v3.SocketAddress ipv4_address_and_port = 3;

  // Port isn't yet available; must be zero or ignored
  config.core.v3.SocketAddress ipv4_dnat_address = 5;
```

Reply:
to make sure I understand, `ipv4_dnat_type` and `ipv6_dnat_type` would go away, and we'd just have:
```
config.core.v3.Address ipv4_dnat_address;
config.core.v3.Address ipv6_dnat_address;
```

is that correct?

Reply:
Yes, that's correct. Plus the addition of `config.core.v3.SocketAddress ipv4_address_and_port` and `config.core.v3.SocketAddress ipv6_address_and_port`

Reply:
Also cc @danzh2010 who originally added this extension

Reply:
That SGTM, thanks @ggreenway !

Reply:
Do we really need port in non-DNAT case? Even in DNAT case, the client doesn't see a different port right? Only the translation unit will translate [SPA:443] to something else [configured_address:port], is this correct?

Reply:
If the DNAT is translating both IP and port, then the client-advertised address may need a different port than the listener.

Also, I'm considering adding support for quic listeners to support `additional_addresses`; if that were true then in the non-DNAT case, the client-advertised address may also need a different port, and the DNAT address may need a different port.

Reply:
The use case is a host in AWS with a single NIC and IP, with both an NLB for initial connection load balancing, and an external EIP for SPA traffic. If the two paths go to different ports, envoy/quiche can easily differentiate which path the traffic is coming from. If they go to the same port, envoy/quiche sees the client attempting SPA as the client trying to do a voluntary migration during the initial handshake.